topi/python/topi/x86/nn.py

```diff
@@ -36,56 +39,180 @@ def schedule_softmax(outs):
     return s
 
 
-@generic.schedule_dense.register(["cpu"])
-def schedule_dense(outs):
-    """Schedule for dense
+@autotvm.register_topi_compute(nn.dense, "cpu", "direct")
+def _declaration_dense(cfg, data, weight, bias=None):
+    batch, _ = get_const_tuple(data.shape)
+
+    # For small batch sizes, don't pack weight into cache-friendly layout
+    # because of overhead in packing and limited reuse from batch dimension
+    # TODO(icemelon9): use a more systematic way to determine which schedule to use
+    if batch <= 16:
```

Comment:
Typically if we need to change the declaration for two methods this is enough of a change that it is a change in strategy (template). We might consider using two different template names here instead (e.g., "direct_pack", "direct_nopack").

Reply:
I agree. But how should we switch between two templates? Should we still keep this function registered for "direct" that switches two templates?

Reply:
Just talked with @icemelon9, it seems that in some cases another schedule template is necessary otherwise there will be great performance loss. Maybe we can register another tag for this template, but also use it in direct method.

Reply:
The issue is that the current approach requires using some static manually determined condition (e.g., a branch on some shape value) to choose the template if we also include it in the direct method. I'm not sure this is a good idea in the long run as that may be brittle and suboptimal depending on the hardware.